### PR TITLE
[SDK-2691] Add PHP 8.1 support

### PR DESCRIPTION
**Pull request by evansims**
### Changes

This PR includes changes necessary to support PHP 8.1, currently in code-freeze and release candidate state and slated for release this November.

- Minor code changes to resolve deprecation warnings.
- Updated unit tests to resolve 8.1 compatibility issues.
- Updated Dockerfiles to properly test 7.4 through 8.1.
- Updated CircleCI configuration to add 8.1 tests.

I've added inline comments to clear up any confusion on individual changes.

### Contributor Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)

### CODE REVIEW

File: src/Auth0.php [hunk 1]
```diff
@@ -611,11 +611,11 @@ public function setAccessTokenExpiration(
      * Get the specified parameter from POST or GET, depending on configured response mode.
      *
      * @param string $parameterName Name of the parameter to pull from the request.
-     * @param int $filter Defaults to FILTER_SANITIZE_STRING. The type of PHP filter_var() filter to apply.
+     * @param int $filter Defaults to \FILTER_UNSAFE_RAW. The type of PHP filter_var() filter to apply.
```
Comment (evansims) [inline src/Auth0.php:614]: PHP's FILTER_SANITIZE_STRING was deprecated in 8.1 in favor of FILTER_UNSAFE_RAW. FILTER_UNSAFE_RAW is PHP's native, default sanitization filter and works largely the same as FILTER_SANITIZE_STRING. [Related Internals discussion.](https://externals.io/message/113657#114079)

File: src/Event/HttpRequestBuilt.php [hunk 2]
```diff
@@ -17,7 +17,7 @@ public function __construct(
         $this->httpRequest = $httpRequest;
     }
 
-    public function &get(): RequestInterface
+    public function get(): RequestInterface
```
Comment (evansims) [inline src/Event/HttpRequestBuilt.php:20]: No need to expressly return these by reference, as PHP inherently passes objects by reference.

File: src/Store/CookieStore.php [hunk 3]
```diff
@@ -200,7 +200,11 @@ public function setState(): self
                     $_COOKIE[$cookieName] = $chunk;
 
                     // Push the updated cookie to the host device for persistence.
-                    setcookie($cookieName, $chunk, $setOptions);
+                    // @codeCoverageIgnoreStart
```
Comment (evansims) [inline src/Store/CookieStore.php:203]: Tests cannot currently run in an isolated process due to a PHPUnit bug, and by extension lack of support within Pest. As such, all our tests that deal with cookies throw warnings because headers will have already been sent by previous tests. Until our unit testing framework fixes process isolation, we'll just skip invoking the cookie issuing methods to avoid triggering these errors.

File: src/Store/SessionStore.php [hunk 4]
```diff
@@ -151,14 +151,18 @@ private function start(): void
...
         if ($sessionId === '' || $sessionId === false) {
-            session_set_cookie_params([
-                'lifetime' => $this->configuration->getCookieExpires(),
-                'domain' => $this->configuration->getCookieDomain(),
-                'path' => $this->configuration->getCookiePath(),
-                'secure' => $this->configuration->getCookieSecure(),
-                'httponly' => true,
-                'samesite' => $this->configuration->getResponseMode() === 'form_post' ? 'None' : 'Lax',
-            ]);
+            // @codeCoverageIgnoreStart
```
Comment (evansims) [inline src/Store/SessionStore.php:154]: See above.

File: src/Store/CookieStore.php [hunk 5]
```diff
@@ -216,7 +220,11 @@ public function setState(): self
 
         foreach ($orphaned as $cookieName) {
             // Push the cookie deletion command to the host device.
-            setcookie($cookieName, '', $deleteOptions);
+            // @codeCoverageIgnoreStart
```
Comment (evansims) [inline src/Store/CookieStore.php:223]: See above.

File: src/Event/HttpResponseReceived.php [hunk 6]
```diff
@@ -21,7 +21,7 @@ public function __construct(
         $this->httpRequest = $httpRequest;
     }
 
-    public function &get(): ResponseInterface
+    public function get(): ResponseInterface
```
Comment (evansims) [inline src/Event/HttpResponseReceived.php:24]: See above.

File: src/Auth0.php [hunk 7]
```diff
@@ -641,7 +641,7 @@ public function getExchangeParameters(): ?object
         if ($code !== null && $state !== null) {
             return (object) [
                 'code' => $code,
-                'state' => $state
+                'state' => $state,
```
Comment (evansims) [inline src/Auth0.php:644]: Unrelated code styling fix.

File: src/Utility/HttpResponsePaginator.php [hunk 8]
```diff
@@ -150,10 +150,13 @@ public function count(): int
...
      * Return the current result at our position, if available.
      *
-     * @return array<mixed>|bool
+     * @return mixed
+     *
+     * @psalm-suppress InvalidAttribute
      *
      * @codeCoverageIgnore
      */
+    #[\ReturnTypeWillChange]
```
Comment (evansims) [inline src/Utility/HttpResponsePaginator.php:159]: With PHP 8.1, an (arguably breaking) change was [implemented based on this RFC](https://wiki.php.net/rfc/internal_method_return_types) which sets "tentative" return types for internal methods. This requires code that overrides those methods to have a matching type. If the type does not match, it will throw a deprecation until PHP 9.

File: tests/Pest.php [hunk 9]
```diff
@@ -1,5 +1,11 @@
 <?php
 
+declare(strict_types=1);
```
Comment (evansims) [inline tests/Pest.php:3]: Importing the code previously in `bootstrap.php` into `Pest.php`, since we've moved to that test framework, and it invokes this script before bootstrap. This ensures AUTH0_TEST_DIR is available earlier in test processing and avoids potential pitfalls when setting up beforeAll and afterAll calls.

File: tests/Unit/API/Management/JobsTest.php [hunk 10]
```diff
@@ -23,7 +23,7 @@
 });
 
 test('createImportUsers() issues an appropriate request', function(): void {
-    $importPath = AUTH0_PHP_TEST_JSON_DIR . 'test-import-users-file.json';
+    $importPath = join(DIRECTORY_SEPARATOR, [AUTH0_TESTS_DIR, 'json', 'test-import-users-file.json']);
```
Comment (evansims) [inline tests/Unit/API/Management/JobsTest.php:26]: `AUTH0_PHP_TEST_JSON_DIR` was an unnecessary global test declaration only used once; just getting rid of that here.

File: src/Utility/Toolkit.php [hunk 11]
```diff
@@ -56,7 +56,7 @@ public static function times(
      * @param \Closure $callback The function to pass each item through.
      */
     public static function each(
-        iterable &$items,
+        iterable & $items,
```
Comment (evansims) [inline src/Utility/Toolkit.php:59]: Styling tweak; fixes 8.1 code style warning in phpcs.

File: src/Utility/HttpRequest.php [hunk 12]
```diff
@@ -138,7 +138,7 @@ public function __construct(
         string $basePath = '/',
         array $headers = [],
         ?string $domain = null,
-        ?array &$mockedResponses = null
+        ?array & $mockedResponses = null
```
Comment (evansims) [inline src/Utility/HttpRequest.php:141]: Styling tweak; fixes 8.1 code style warning in phpcs.

File: src/Token/Verifier.php [hunk 13]
```diff
@@ -93,7 +93,7 @@ public function __construct(
         ?string $clientSecret = null,
         ?int $cacheExpires = null,
         ?CacheItemPoolInterface $cache = null,
-        ?array &$mockedHttpResponses = null
+        ?array & $mockedHttpResponses = null
```
Comment (evansims) [inline src/Token/Verifier.php:96]: Styling tweak; fixes 8.1 code style warning in phpcs.

File: .circleci/config.yml [hunk 14]
```diff
@@ -130,7 +130,7 @@ commands:
     steps:
       - run:
           name: Run unit tests (Pest)
-          command: php -d memory_limit=2G ./vendor/bin/pest --stop-on-failure --min=80 --coverage-clover=<< parameters.clover >>/coverage.xml --coverage-xml=<< parameters.clover >>/coverage-xml --log-junit=<< parameters.clover >>/junit.xml
+          command: php -d error_reporting='E_ALL ^ E_DEPRECATED' -d session.use_cookies=false -d session.cache_limiter=false -d memory_limit=2G ./vendor/bin/pest --stop-on-failure --min=95 --coverage-clover=<< parameters.clover >>/coverage.xml --coverage-xml=<< parameters.clover >>/coverage-xml --log-junit=<< parameters.clover >>/junit.xml
```
Comment (evansims) [inline .circleci/config.yml:133]: This ensures test conditions are set satisfactorily, rather than having to mutate php.inis or rely on bootstrap.php conditions on the host device: pass them from the CLI directly.

File: .circleci/config.yml [hunk 15]
```diff
@@ -205,7 +205,6 @@ workflows:
       - run-tests:
           name: run-tests-8.0
           php: "8.0.9"
-      # - run-tests:
-      #     name: run-tests-8.1
-      #     php: "8.1.0beta1"
-      #     composer: "--ignore-platform-reqs"
+      - run-tests:
```
Comment (evansims) [inline .circleci/config.yml:208]: Adds 8.1 to test suite on CircleCI.

File: composer.json [hunk 16]
```diff
@@ -28,7 +28,7 @@
     }
   ],
   "require": {
-    "php": "^7.4 || ^8.0",
+    "php": "^7.4 || ^8.0, <8.2",
```
Comment (evansims) [inline composer.json:31]: We support PHP 7.4, 8.0 and 8.1. Don't suggest we support 8.2, which will be entering development shortly, until we've tested that.

File: composer.json [hunk 17]
```diff
@@ -50,14 +50,14 @@
     "mockery/mockery": "^1.4",
     "nunomaduro/phpinsights": "^2.0",
     "nyholm/psr7": "^1.4",
-    "pestphp/pest": "^1.0",
-    "pestphp/pest-plugin-parallel": "^0.2.0",
+    "pestphp/pest": "^1.18",
```
Comment (evansims) [inline composer.json:53]: Ensure we're seeding earliest compatible versions of Pest and Psalm for the new 8.1 tests.

File: composer.json [hunk 18]
```diff
@@ -74,17 +74,43 @@
...
       "@tests:phpstan",
       "@tests:psalm",
       "@tests:pest"
     ],
-    "tests:phpinsights": "@php ./vendor/bin/phpinsights -v --no-interaction",
-    "tests:phpstan": "@php ./vendor/bin/phpstan analyse --ansi --memory-limit 512M",
-    "tests:psalm": "@php ./vendor/bin/psalm",
-    "tests:pest": "@php ./vendor/bin/pest --stop-on-failure --coverage",
-    "tests:infection": "@php ./vendor/bin/infection --threads=4 --test-framework=pest"
+    "tests:phpinsights": [
```
Comment (evansims) [inline composer.json:83]: - Disable Composer's process timeout for scripts.
- Add new Dockerfile tests for 7.4, 8.0 and 8.1.

File: docker/php-8.1.dockerfile [hunk 19]
```diff
@@ -0,0 +1,23 @@
+FROM php:8.1.0RC1-cli-alpine
```
Comment (evansims) [inline docker/php-8.1.dockerfile:1]: Dockerfile for PHP 8.1RC1 tests.

File: docker/php-8.0.dockerfile [hunk 20]
```diff
@@ -0,0 +1,23 @@
+FROM php:8.0-cli-alpine
```
Comment (evansims) [inline docker/php-8.0.dockerfile:1]: Dockerfile for PHP 8.0 tests.

File: docker/php-7.4.dockerfile [hunk 21]
```diff
@@ -0,0 +1,23 @@
+FROM php:7.4-cli-alpine
```
Comment (evansims) [inline docker/php-7.4.dockerfile:1]: Dockerfile for PHP 7.4 tests.

File: docker/entrypoint.sh [hunk 22]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
```
Comment (evansims) [inline docker/entrypoint.sh:1]: ENTRYPOINT script for all Dockerfile tests. Validates composer.json, installs dependencies, and runs our test suite. Finally echos PHP version for tracking purposes.

File: docker-compose.yml [hunk 23]
```diff
@@ -1,16 +1,30 @@
 version: "3.9"
 
 services:
-  tests:
-    image: composer:2.1
+  tests-7.4:
```
Comment (evansims) [inline docker-compose.yml:4]: Add new Dockfiles to our docker-compose manifest for more robust containerized tests, and support for 8.1 on devices where executables may not be ready for testing yet.

File: phpunit.xml.dist [hunk 24]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0"?>
-<phpunit bootstrap="./tests/bootstrap.php" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" colors="true" verbose="true" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" stderr="true" colors="true" verbose="true" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
```
Comment (evansims) [inline phpunit.xml.dist:2]: Remove bootstrap file in favor of Pest.php (covered below) and ensure we're properly capturing stderr output.

File: .circleci/config.yml (hunk 15 above)
Comment (lbalmaceda) [reply .circleci/config.yml:208]: Consider using [matrix jobs](https://circleci.com/blog/circleci-matrix-jobs/) if you are going to keep adding different versions for the same job.

File: .circleci/config.yml (hunk 14 above)
Comment (lbalmaceda) [reply .circleci/config.yml:133]: Is this line enforcing the coverage? If true, shouldn't that be enforced as part of the codecov configuration?
I noticed this repo is not configured https://about.codecov.io/language/php/

File: docker/entrypoint.sh [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+
+composer install --no-interaction --prefer-dist
+
+php ./vendor/bin/phpinsights -v --no-interaction
+
+php ./vendor/bin/phpstan analyse --ansi --memory-limit 512M
+
+php ./vendor/bin/psalm
+
+php -d error_reporting='E_ALL ^ E_DEPRECATED' -d session.use_cookies=false -d session.cache_limiter=false ./vendor/bin/pest --stop-on-failure --coverage
```
Comment (lbalmaceda) [inline docker/entrypoint.sh:13]: I'm curious, I would expect this line to match the line you are using in the CI configuration but I see some differences. Is that accurate?

File: src/Event/HttpRequestBuilt.php (hunk 2 above)
Comment (lbalmaceda) [reply src/Event/HttpRequestBuilt.php:20]: would this be a breaking change?

**Review by lbalmaceda — APPROVED** [review]
Left a few comments, not blockers.

File: src/Event/HttpRequestBuilt.php (hunk 2 above)
Comment (evansims) [reply src/Event/HttpRequestBuilt.php:20]: Nope! The result is the same and compatible with any previous iterations (although this class is new to 8.0, anyway.) It was just a redundant language instruction; & tells PHP to pass something by reference rather than value, such as an array value; but objects, such as this case, are always passed by reference, so it was just unnecessary.

File: .circleci/config.yml (hunk 14 above)
Comment (evansims) [reply .circleci/config.yml:133]: For our PHP repos I opted to use PHP's own code coverage tooling as part of our test suites (pcov) to keep us informed on those coverage metrics, rather than relying on codecov. Just not a fan of it personally, and pcov works really smoothly for us on these PHP repos.

But this particular change is just for silencing deprecation warnings from upstream library dependencies we use that haven't fully prepared for PHP 8.1 yet; it gets very noisy unless we silence those. I'll remove them once our upstreams get updated with PHP 8.1 support and our tests are readable again without it all.

File: .circleci/config.yml (hunk 15 above)
Comment (evansims) [reply .circleci/config.yml:208]: Didn't know about those, cool! I think I effectively reinvented that feature with my CircleCI config using variables, haha. I'll work on shifting over to those as time allows, much cleaner.

File: docker/entrypoint.sh (hunk 25 above)
Comment (evansims) [reply docker/entrypoint.sh:13]: Good catch; the Docker images, as well as the local tests, run PEST/PHPUnit in a slightly different configuration. CircleCI includes extra configuration to store the coverage data for the mutation testing to pick up in a separate task, otherwise the mutation tests would need to run the unit tests a second time to get that data itself when the task runs. It's just caching that data for CircleCI purposes, basically, which isn't necessary on local or Docker testing.

**Comment by evansims** [comment]
Thanks for the thorough review, @lbalmaceda! I replied to those questions to help clarify 👍
